Tallies/TallyMaps/cellMap_class.f90

```diff
@@ -0,0 +1,272 @@
+    integer(shortInt)                             :: Nbins   = 0
+    integer(shortInt), dimension(:), allocatable  :: cellIndices
+
+  contains
+    ! Superclass interface implementaction
+    procedure :: init
+    procedure :: bins
+    procedure :: map
+    procedure :: getAxisName
+    procedure :: print
+    procedure :: kill
+
+    ! Class specific procedures
+    procedure :: build
+
+  end type cellMap
+
+contains
+
+  !!
+  !! Build cell map from components directly
+  !!
+  !! Args:
+  !!   cells [in]     -> Array of cell IDs to be included in the map
+  !!   trackRest [in] -> Logical. For TRUE makes extra bin for all cells not
+  !!     explicitly in the map.
+  !!
+  !! Erorrs:
+  !!   None from here. Geometry should give fatalError if a cell is given but
+  !!   not defined.
+  !!
+  subroutine build(self, cells, trackRest)
+    class(cellMap), intent(inout)               :: self
+    integer(shortInt), dimension(:), intent(in) :: cells
+    logical(defBool), intent(in)                :: trackRest
+    integer(shortInt)                           :: N, i, ID
+    class(geometry), pointer                    :: geom
+    class(geometryStd), pointer                 :: geomStd
+
+    ! Find number of cells to bin
+    N = size(cells)
+
+    ! Allocate space in map and cellIndices
+    call self % binMap % init(N)
+    allocate(self % cellIndices(N))
+
+    ! Get pointer to geometry
+    ! ASSUMES THERE IS ONLY ONE GEOMETRY, AND THAT IT IS STD
+    geom => geomPtr(1)
+    geomStd => geometryStd_CptrCast(geom)
+
+    ! Load cell indices and bins
+    do i = 1,N
+      ! Get unique cell IDs
+      ID = geomStd % geom % cells % getIdx(cells(i))
+      call self % binMap % add(ID, i)
+      self % cellIndices(i) = cells(i)
+    end do
+
+    ! Set default and number of bins
+    if(trackRest) then
+      self % Nbins   = N + 1
+      self % default = N + 1
+
+    else
+      self % Nbins   = N
+      self % default = 0
+    end if
+
+  end subroutine build
+
+  !!
+  !! Initialise cell map from dictionary
+  !!
+  !! See tallyMap for specification
+  !!
+  subroutine init(self, dict)
+    class(cellMap), intent(inout)              :: self
+    class(dictionary), intent(in)              :: dict
+    integer(shortInt),dimension(:),allocatable :: cellIDs
+    character(nameLen)                         :: undefined
+    logical(defBool)                           :: trackUndefined
+    character(100), parameter :: Here = 'init (cellMap_class.f90)'
+
+    ! Get cell names list
+    call dict % get(cellIDs, 'cells')
+
+    ! Get setting for undefined tracking
+    call dict % getOrDefault(undefined, 'undefBin', 'false')
+
+    select case (undefined)
+      case('yes','y','true','TRUE','T')
+        trackUndefined = .true.
+
+      case('no', 'n', 'false', 'FALSE', 'F')
+        trackUndefined = .false.
+
+      case default
+        call fatalError(Here, undefined//' is an unrecognised entry!')
+    end select
```

Comment:
We really need to make a function for that conversion. Not in this PR but I guess we should start an issue.